File: javascript/core/src/closure.js
// Closure example in JavaScript
// A closure is a function that retains access to its lexical scope, even when the function is executed outside that scope.
// Closures are often used to create private variables or functions, allowing for encapsulation and data hiding.
// like #private variables in other languages, but in JavaScript, we can achieve this using closures.
// #filed is now alternatively used to create private variables in JavaScript.
function createCounter() {
  let count = 0; // This variable is enclosed

  return function () {
    count++;
    return count;
  };
}

const counter = createCounter(); // 'count' is now preserved in closure
console.log(counter()); // 1
console.log(counter()); // 2

console.log(counter().count); // undefined, private

File: perl/core/bin/closure.pl
#!/usr/bin/perl
use strict;
use warnings;

# This script demonstrates how to create a closure in Perl.
# A closure is a subroutine that captures the lexical variables from its surrounding scope.
# It allows the subroutine to maintain state between calls.
# This example creates a simple counter using a closure.
# its purpose is to demonstrate how closures can be used to encapsulate state. specially, in a class-like manner.
sub make_counter {
    my $count = 0;

    return sub {
        $count++;
        return $count;
    };
}

my $counter1 = make_counter();
print $counter1->();  # prints 1
print $counter1->();  # prints 2

my $counter2 = make_counter();
print $counter2->();  # prints 1 (separate state)
